python/ray/tune/utils/placement_groups.py

```diff
@@ -0,0 +1,188 @@
+from collections import defaultdict
+from typing import Dict, Optional, Set, Tuple
+import os
+import time
+
+import ray
+from ray import ObjectRef
+from ray.actor import ActorClass
+from ray.tune.resources import PlacementGroupFactory
+from ray.tune.trial import Trial
+from ray.util.placement_group import PlacementGroup
+
+TUNE_MAX_PENDING_TRIALS_PG = int(os.getenv("TUNE_MAX_PENDING_TRIALS_PG", 1000))
+# Seconds we wait for a trial to come up before we make blocking calls
+# to process events
+TUNE_TRIAL_STARTUP_GRACE_PERIOD = float(
+    os.getenv("TUNE_TRIAL_STARTUP_GRACE_PERIOD", 10.))
+
+
+class PlacementGroupManager:
+    """PlacementGroupManager to stage and manage placement groups.
+
+    This class schedules placement groups for trials, keeps track of
+    their state, and can return a fully configured actor class using
+    this placement group.
+
+    If two trials share the same placement group factory, both could use
+    resulting placement groups from it. Thus this manager associates
+    placement groups with their factory methods.
+    """
+
+    def __init__(self):
+        # Sets of staged placement groups by factory
+        self._staging: Dict[PlacementGroupFactory, Set[
+            PlacementGroup]] = defaultdict(set)
+        # Sets of ready and unused placement groups by factory
+        self._ready: Dict[PlacementGroupFactory, Set[
+            PlacementGroup]] = defaultdict(set)
+        # Ray futures to check if a placement group is ready
+        self._staging_futures: Dict[ObjectRef, Tuple[PlacementGroupFactory,
+                                                     PlacementGroup]] = {}
+
+        # Placement groups used by trials
+        self._in_use_pgs: Dict[PlacementGroup, Trial] = {}
+        self._in_use_trials: Dict[Trial, PlacementGroup] = {}
+
+        # Latest PG staging time to check if still in grace period.
+        self._latest_staging_start_time = time.time()
+
+    def stage_trial_pg(self, pgf: PlacementGroupFactory):
+        """Stage a trial placement group.
+
+        Create the trial placement group if maximum number of pending
+        placement groups is not exhausted.
+
+        Args:
+            pgf (PlacementGroupFactory): Placement group factory to stage.
+
+        Returns:
+            False if placement group has not been staged, True otherwise.
+
+        Creates placement group and moves it to `self._staging`.
+        """
+        if not self.can_stage():
+            return False
+
+        pg = pgf()  # This creates the placement group
+
+        self._staging[pgf].add(pg)
```

Comment:
Hmm, this is nice -- one thing I could foresee is that if you do experiment stop / resume, you could end up serializing the trials in a way that they all have their own copy of the pgf. Then, when you deserialize, you end up with 100 different pgfs (one object for each trial). Perhaps there's some way to guarantee unique hashing (but this isn't a blocker)

Reply:
Yes that'll be a problem - but it would have been the same for the PG switching. A way we could avoid this would be by having the placement group factories return not placement groups, but bundle definitions and optional kwargs (e.g. for the `strategy` parameter). The benefit here would be that we have insights into the PG definition without instantiating one. We can then use this for hashing/comparing factories, but also to gain information on the requested resources (e.g. for head resource management).